### ISSUE TITLE
segfault after upgrade to 1.6.5-2

### ISSUE THREAD

**Issue by ebal** [vstakhov/rspamd#1956]
### Classification (Please choose *one* option):

* [x] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [ ] Other bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.6.5-2

### Operation system, CPU, memory and environment:
CentOS release 6.9
Linux kvm 2.6.32-696.16.1.el6.centos.plus.x86_64

### Description (Please provide a descriptive summary of the issue):
I've just upgrade rspamd from 1.6.5-1 to 1.6.5-2 

Oct 23 14:02:46 Updated: rspamd-1.6.5-1.x86_64
Dec 10 23:34:15 Updated: rspamd-1.6.5-2.x86_64

and having these segfault issues:

Dec 10 23:53:25 kvm kernel: rspamd[2969]: segfault at 69423088 ip 00007f426eddd109 sp 00007fff291625f0 error 4 in libglib-2.0.so.0.2800.8[7f426edc6000+115000]

### Additional information:

Tried to reinstall glib2-2.28.8-9.el6.x86_64, rspamd-1.6.5-2.x86_64 but nothing helped

**Comment by vstakhov** [comment]
https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed

**Comment by ebal** [comment]
Reading symbols from /usr/bin/rspamd...Reading symbols from /usr/lib/debug/usr/bin/rspamd.debug...done.
done.

warning: core file may not match specified executable file.
[New Thread 4933]
Reading symbols from /lib64/ld-linux-x86-64.so.2...Reading symbols from /usr/lib/debug/lib64/ld-2.12.so.debug...done.
done.
Loaded symbols for /lib64/ld-linux-x86-64.so.2
Core was generated by `/usr/bin/rspamadm configtest -c /etc/rspamd/rspamd.sysvinit.conf'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007fc35410d109 in ?? ()
(gdb)

**Comment by ebal** [comment]
[rspamadm.core.4933.gz](https://github.com/vstakhov/rspamd/files/1546155/rspamadm.core.4933.gz)

**Comment by vstakhov** [comment]
You need symbols from `rspamadm` here:

```
Reading symbols from /usr/bin/rspamd...Reading symbols from /usr/lib/debug/usr/bin/rspamd.debug...done.
done.

warning: core file may not match specified executable file.
```

Try it with `gdb /usr/bin/rspamadm <core_file>`

**Comment by ebal** [comment]
Here it is, I hope it helps.
I've also uploaded the core file to the previous comment.
Thanks,


`
[root@kvm coreland]# gdb /usr/bin/rspamadm rspamadm.core.4933 
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-92.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/bin/rspamadm...Reading symbols from /usr/lib/debug/usr/bin/rspamadm.debug...done.
done.
[New Thread 4933]
Reading symbols from /usr/lib64/libjemalloc.so.1...Reading symbols from /usr/lib/debug/usr/lib64/libjemalloc.so.1.debug...done.
done.
Loaded symbols for /usr/lib64/libjemalloc.so.1
Reading symbols from /lib64/libpthread.so.0...Reading symbols from /usr/lib/debug/lib64/libpthread-2.12.so.debug...done.
[Thread debugging using libthread_db enabled]
done.
Loaded symbols for /lib64/libpthread.so.0
Reading symbols from /lib64/libgthread-2.0.so.0...Reading symbols from /usr/lib/debug/lib64/libgthread-2.0.so.0.2800.8.debug...done.
done.
Loaded symbols for /lib64/libgthread-2.0.so.0
Reading symbols from /lib64/librt.so.1...Reading symbols from /usr/lib/debug/lib64/librt-2.12.so.debug...done.
done.
Loaded symbols for /lib64/librt.so.1
Reading symbols from /lib64/libglib-2.0.so.0...Reading symbols from /usr/lib/debug/lib64/libglib-2.0.so.0.2800.8.debug...done.
done.
Loaded symbols for /lib64/libglib-2.0.so.0
Reading symbols from /lib64/libpcre.so.0...Reading symbols from /usr/lib/debug/lib64/libpcre.so.0.0.1.debug...done.
done.
Loaded symbols for /lib64/libpcre.so.0
Reading symbols from /usr/lib64/libevent-1.4.so.2...Reading symbols from /usr/lib/debug/usr/lib64/libevent-1.4.so.2.1.3.debug...done.
done.
Loaded symbols for /usr/lib64/libevent-1.4.so.2
Reading symbols from /usr/lib64/libsqlite3.so.0...Reading symbols from /usr/lib/debug/usr/lib64/libsqlite3.so.0.8.6.debug...done.
done.
Loaded symbols for /usr/lib64/libsqlite3.so.0
Reading symbols from /usr/lib64/libicudata.so.42...Reading symbols from /usr/lib/debug/usr/lib64/libicudata.so.42.1.debug...
warning: "/usr/lib/debug/usr/lib64/libicudata.so.42.1.debug": separate debug info file has no debug info
(no debugging symbols found)...done.
(no debugging symbols found)...done.
Loaded symbols for /usr/lib64/libicudata.so.42
Reading symbols from /usr/lib64/libicuuc.so.42...Reading symbols from /usr/lib/debug/usr/lib64/libicuuc.so.42.1.debug...done.
done.
Loaded symbols for /usr/lib64/libicuuc.so.42
Reading symbols from /usr/lib64/libicui18n.so.42...Reading symbols from /usr/lib/debug/usr/lib64/libicui18n.so.42.1.debug...done.
done.
Loaded symbols for /usr/lib64/libicui18n.so.42
Reading symbols from /usr/lib64/libssl.so.10...Reading symbols from /usr/lib/debug/usr/lib64/libssl.so.1.0.1e.debug...done.
done.
Loaded symbols for /usr/lib64/libssl.so.10
Reading symbols from /usr/lib64/libcrypto.so.10...Reading symbols from /usr/lib/debug/usr/lib64/libcrypto.so.1.0.1e.debug...done.
done.
Loaded symbols for /usr/lib64/libcrypto.so.10
Reading symbols from /usr/lib64/libmagic.so.1...Reading symbols from /usr/lib/debug/usr/lib64/libmagic.so.1.0.0.debug...done.
done.
Loaded symbols for /usr/lib64/libmagic.so.1
Reading symbols from /lib64/libm.so.6...Reading symbols from /usr/lib/debug/lib64/libm-2.12.so.debug...done.
done.
Loaded symbols for /lib64/libm.so.6
Reading symbols from /usr/lib64/libfann.so.2...Reading symbols from /usr/lib/debug/usr/lib64/libfann.so.2.2.0.debug...done.
done.
Loaded symbols for /usr/lib64/libfann.so.2
Reading symbols from /lib64/libdl.so.2...Reading symbols from /usr/lib/debug/lib64/libdl-2.12.so.debug...done.
done.
Loaded symbols for /lib64/libdl.so.2
Reading symbols from /lib64/libresolv.so.2...Reading symbols from /usr/lib/debug/lib64/libresolv-2.12.so.debug...done.
done.
Loaded symbols for /lib64/libresolv.so.2
Reading symbols from /lib64/libnsl.so.1...Reading symbols from /usr/lib/debug/lib64/libnsl-2.12.so.debug...done.
done.
Loaded symbols for /lib64/libnsl.so.1
Reading symbols from /usr/lib64/rspamd/librspamd-actrie.so...Reading symbols from /usr/lib/debug/usr/lib64/rspamd/librspamd-actrie.so.debug...done.
done.
Loaded symbols for /usr/lib64/rspamd/librspamd-actrie.so
Reading symbols from /usr/lib64/liblua-5.1.so...(no debugging symbols found)...done.
Loaded symbols for /usr/lib64/liblua-5.1.so
Reading symbols from /lib64/libc.so.6...Reading symbols from /usr/lib/debug/lib64/libc-2.12.so.debug...done.
done.
Loaded symbols for /lib64/libc.so.6
Reading symbols from /lib64/ld-linux-x86-64.so.2...Reading symbols from /usr/lib/debug/lib64/ld-2.12.so.debug...done.
done.
Loaded symbols for /lib64/ld-linux-x86-64.so.2
Reading symbols from /usr/lib64/libstdc++.so.6...Reading symbols from /usr/lib/debug/usr/lib64/libstdc++.so.6.0.13.debug...done.
done.
Loaded symbols for /usr/lib64/libstdc++.so.6
Reading symbols from /lib64/libgcc_s.so.1...Reading symbols from /usr/lib/debug/lib64/libgcc_s-4.4.7-20120601.so.1.debug...done.
done.
Loaded symbols for /lib64/libgcc_s.so.1
Reading symbols from /lib64/libgssapi_krb5.so.2...Reading symbols from /usr/lib/debug/lib64/libgssapi_krb5.so.2.2.debug...done.
done.
Loaded symbols for /lib64/libgssapi_krb5.so.2
Reading symbols from /lib64/libkrb5.so.3...Reading symbols from /usr/lib/debug/lib64/libkrb5.so.3.3.debug...done.
done.
Loaded symbols for /lib64/libkrb5.so.3
Reading symbols from /lib64/libcom_err.so.2...Reading symbols from /usr/lib/debug/lib64/libcom_err.so.2.1.debug...done.
done.
Loaded symbols for /lib64/libcom_err.so.2
Reading symbols from /lib64/libk5crypto.so.3...Reading symbols from /usr/lib/debug/lib64/libk5crypto.so.3.1.debug...done.
done.
Loaded symbols for /lib64/libk5crypto.so.3
Reading symbols from /lib64/libz.so.1...Reading symbols from /usr/lib/debug/lib64/libz.so.1.2.3.debug...done.
done.
Loaded symbols for /lib64/libz.so.1
Reading symbols from /lib64/libkrb5support.so.0...Reading symbols from /usr/lib/debug/lib64/libkrb5support.so.0.1.debug...done.
done.
Loaded symbols for /lib64/libkrb5support.so.0
Reading symbols from /lib64/libkeyutils.so.1...Reading symbols from /usr/lib/debug/lib64/libkeyutils.so.1.3.debug...done.
done.
Loaded symbols for /lib64/libkeyutils.so.1
Reading symbols from /lib64/libselinux.so.1...Reading symbols from /usr/lib/debug/lib64/libselinux.so.1.debug...done.
done.
Loaded symbols for /lib64/libselinux.so.1
Core was generated by `/usr/bin/rspamadm configtest -c /etc/rspamd/rspamd.sysvinit.conf'.
Program terminated with signal 11, Segmentation fault.
#0  g_ptr_array_maybe_expand (array=0x4e9c2600, len=1) at garray.c:964
964	  if ((array->len + len) > array->alloc)

`

**Comment by vstakhov** [comment]
Can you run `bt full` in gdb using this core?

**Comment by ebal** [comment]
attached 
[core.txt.zip](https://github.com/vstakhov/rspamd/files/1547294/core.txt.zip)
thanks

**Comment by GrCat** [comment]
I guess we have a similar problem.
We have upgrade from 1.6.5-1 to 1.6.5-2.
CentOS release 6.9 (Final), 2.6.32-696.16.1.el6.x86_64 #1 SMP

rspamd[1772]: segfault at ffffffffb5023328 ip 00007f66ba854109 sp 00007ffe5413aa50 error 4 in libglib-2.0.so.0.2800.8[7f66ba83d000+115000]

rspamadm exit with segmentation fault.
I will be grateful for your help

**Comment by vstakhov** [comment]
I think it is CentOS issue. It has appeared when I have enabled jemalloc in build. Does it crash when you set `G_SLICE=always-malloc` environment variable? I'm sorry, but I have no time to investigate this issue so far.

**Comment by GrCat** [comment]
Unfortunately, it continues to crash.
Can the temporary solution be a rollback to version 1.6.5-1? But I can not find such rpms

**Comment by vstakhov** [comment]
Yes, old rpms are removed.

**Comment by ebal** [comment]
Hi, I am not quite sure if this help,
I was looking on your rpm spec files trying to build rspamd into a "fresh/clean" centos6 docker image.

below are my personal notes: 
https://github.com/ebal/SPECS/blob/master/rspamd.notes
https://github.com/ebal/SPECS/blob/master/rspamd.spec

seems to be working for me, but I can not give any advice on the matter as I do not have much experience in building rpm files.  Hopefully when you have some time, you can take a look and  test this.

Thanks in advance,

**Comment by christiananthon** [comment]
It would be good to have this sorted. Setting G_SPLICE=always-malloc doesn't seem to help, dowgrading does. @vstakhov  Could you provide the spec file you are currently using?

**Comment by christiananthon** [comment]
I took a look at the code, and the change doesn't seem benign, in the sense that the code does something different even with jemalloc disabled.

**Comment by christiananthon** [comment]
I can confirm that there is no crash on startup in centos-6 in my own build. However, enabling jemalloc introduces the problem. Maybe jemalloc can be disabled for el6 and friends?

**Comment by fatalbanana** [comment]
I think that's what's happened. You could try 1.6.5-3.

**Comment by christiananthon** [comment]
Nope, same problem
[root@dna01 ~]# service rspamd restart
/etc/init.d/rspamd: line 91: 114311 Segmentation fault      (core dumped) $rspamadm configtest -c $RSPAMD_CONF_FILE
[root@dna01 ~]# ldd /usr/bin/rspamd | grep jem
        libjemalloc.so.1 => /usr/lib64/libjemalloc.so.1 (0x00000034aec00000)
[root@dna01 ~]# rpmquery -a | grep rspamd
rspamd-1.6.5-3.x86_64

**Comment by christiananthon** [comment]
Compared to 
[root@dna01 ~]# service rspamd restart
syntax OK
Stopping rspamd.                                           [  OK  ]
Starting rspamd:                                           [  OK  ]
[root@dna01 ~]# rpmquery -a | grep rspamd
rspamd-1.6.5-1.x86_64
[root@dna01 ~]# ldd /usr/bin/rspamd | grep jem
[root@dna01 ~]#

**Comment by fatalbanana** [comment]
Should be fixed in 1.6.5-4 in fact. There are positive indications over at #1978
